x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/RandomizedAssignmentRounding.java

```diff
@@ -0,0 +1,419 @@
+                        }
+                    }
+                    assignExcessCores(n);
+                }
+            }
+        }
+
+        private void assignModelToNode(Model m, Node n, int allocations) {
+            Tuple<Model, Node> assignment = Tuple.tuple(m, n);
+            int assignedAllocations = Math.min(allocations, resourceTracker.remainingModelAllocations.get(m));
+            softAssignments.put(assignment, 1.0);
+            softAllocations.put(assignment, (double) assignedAllocations);
+            resourceTracker.assign(m, n, assignedAllocations);
+        }
+
+        private double decreasingQualityNodeOrder(Node n) {
+            double quality = 0.0;
+            for (Model m : models) {
+                Tuple<Model, Node> index = Tuple.tuple(m, n);
+                if (softAllocations.get(index) > 0) {
+                    quality += (1 + (m.currentAllocationByNodeId().containsKey(n.id()) ? 1 : 0)) * softAllocations.get(index) * m
+                        .threadsPerAllocation();
+                }
+            }
+            return quality;
+        }
+
+        private void assignExcessCores(Node n) {
+            if (resourceTracker.remainingNodeCores.get(n) == 0) {
+                return;
+            }
+
+            if (hasSoftAssignments(n)) {
+                return;
+            }
+
+            // We know the models on this node are definitely assigned thus we can also
+            // assign any extra cores this node has to the models in descending size order.
+            for (Model m : models.stream()
+                .filter(m -> softAssignments.get(Tuple.tuple(m, n)) == 1 && resourceTracker.remainingModelAllocations.get(m) > 0)
+                .sorted(Comparator.comparingDouble(this::remainingModelOrder))
+                .toList()) {
+                if (resourceTracker.remainingNodeCores.get(n) <= 0) {
+                    break;
+                }
+                int extraAllocations = Math.min(
+                    resourceTracker.remainingNodeCores.get(n) / m.threadsPerAllocation(),
+                    resourceTracker.remainingModelAllocations.get(m)
+                );
+                softAllocations.compute(Tuple.tuple(m, n), (k, v) -> v + extraAllocations);
+                resourceTracker.assign(m, n, extraAllocations);
+            }
+
+            zeroSoftAssignmentsOfSatisfiedModels();
+        }
+
+        private double remainingModelOrder(Model m) {
+            return (m.currentAllocationByNodeId().isEmpty() ? 1 : 2) * -m.memoryBytes();
+        }
+
+        private boolean hasSoftAssignments(Node n) {
+            return models.stream().anyMatch(m -> isSoftAssignment(m, n));
+        }
+
+        private boolean isSoftAssignment(Model m, Node n) {
+            Tuple<Model, Node> index = Tuple.tuple(m, n);
+            return (softAssignments.get(index) > 0 && softAssignments.get(index) < 1) || isInteger(softAllocations.get(index)) == false;
```

Comment:
shouldn't this also do a `null` check?

Reply:
We initialize `softAssignments` so that it contains an entry for every possible assignment (m, n).

Reply:
`softAssignments` could be a `m * n` array of doubles index by the model and node order avoiding the use of the Tuple to key the map entry.

Reply:
`softAssignments` could be a `m * n` array of doubles index by the model and node order avoiding the use of the Tuple to key the map entry.

Reply:
But at places we need to access them having a `(m, n)` pair. How would we convert those to the right indices?

Reply:
store m and n in separate maps to an int the int is the index in the double array. I am not sure this is better or not.